components/director/internal/domain/apptemplate/service.go

```diff
@@ -338,9 +338,61 @@ func (s *service) PrepareApplicationCreateInputJSON(appTemplate *model.Applicati
 
 		appCreateInputJSON = strings.ReplaceAll(appCreateInputJSON, fmt.Sprintf("{{%s}}", placeholder.Name), newValue)
 	}
+
+	for _, placeholder := range appTemplate.Placeholders {
+		newValue, err := values.FindPlaceholderValue(placeholder.Name)
+		isOptional := false
+		if placeholder.Optional != nil {
+			isOptional = *placeholder.Optional
+		}
+
+		if err != nil && !isOptional {
+			return "", errors.Wrap(err, "required placeholder not provided")
+		}
```

Comment:
I see that this logic is the same as the one from l323 to l332. Can we somehow do the whole thing in one loop? If not so, perhaps consider taking it out in a separate function.

Reply:
Not possible to be done on one pass.
The idea of first pass is to process entries and the second pass require the appCreateInputJSON to be already filled and now we just clear the empty labels from it.

Reply:
Now I clear the output now on each property process. Slower than before, but penalty can be ignored, as data that is processed is usually les than 15 fields.